Mathlib/GroupTheory/GroupAction/MultipleTransitivity.lean

```diff
@@ -0,0 +1,353 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.GroupTheory.GroupAction.Embedding
+import Mathlib.GroupTheory.GroupAction.Primitive
+import Mathlib.GroupTheory.GroupAction.Transitive
+
+import Mathlib.Data.Fin.Tuple.Embedding
+
+/-! # Multiple transitivity
+
+* `MulAction.IsMultiplyPretransitive`:
+  A multiplicative action of a group `G` on a type `α` is n-transitive
+  if the action of `G` on `Fin n ↪ α` is pretransitive.
+
+* Any action is 0-pretransitive
+
+* `MulAction.is_one_pretransitive_iff` :
+  An action is 1-pretransitive iff it is pretransitive
+
+* `MulAction.is_two_pretransitive_iff` :
+  An action is 2-pretransitive if for any `a`, `b`, `c`, `d`, such that
+  `a ≠ b` and `c ≠ d`, there exist `g : G` such that `g • a = b` and `g • c = d`.
+
+* `MulAction.isPreprimitive_of_is_two_pretransitive` :
+  A 2-transitive action is preprimitive
+
+* `MulAction.isMultiplyPretransitive_of_le` :
+  If an action is `n`-pretransitive, then it is `m`-pretransitive for all `m ≤ n`,
+  provided `α` has at least `n` elements.
+
+## Remarks on implementation
+
+These results are results about actions on types `n ↪ α` induced by an action
+on `α`, and some results are developed in this context.
+
+-/
+
+open MulAction MulActionHom Function.Embedding Fin Set Nat
+
+namespace Fin.Embedding
+
+variable {α : Type*}
+
+theorem exists_embedding_disjoint_range_of_add_le_ENat_card
+    {s : Set α} [Finite s] {n : ℕ} (hs : s.ncard + n ≤ ENat.card α) :
+    ∃ y : Fin n ↪ α, Disjoint s (range y) := by
+  suffices Nonempty (Fin n ↪ (sᶜ : Set α)) by
+    obtain ⟨y⟩ := this
+    use y.trans (subtype _)
+    rw [Set.disjoint_right]
+    rintro _ ⟨i, rfl⟩
+    simp only [trans_apply, subtype_apply, ← mem_compl_iff]
+    exact Subtype.coe_prop (y i)
+  rcases finite_or_infinite α with hα | hα
+  · have _ : Fintype α := Fintype.ofFinite α
+    classical
+    apply Function.Embedding.nonempty_of_card_le
+    rwa [Fintype.card_fin, ← add_le_add_iff_left s.ncard,
+      ← Nat.card_eq_fintype_card, Set.Nat.card_coe_set_eq,
+        Set.ncard_add_ncard_compl, ← ENat.coe_le_coe,
+        ← ENat.card_eq_coe_natCard, ENat.coe_add]
+  · exact ⟨valEmbedding.trans s.toFinite.infinite_compl.to_subtype.natEmbedding⟩
+
+theorem restrictSurjective_of_add_le_ENatCard
```

Comment:
Can you move this to a more basic file?

Reply:
It's not clear which. The `#min_imports` says I need
```
import Mathlib.Data.ENat.Lattice
import Mathlib.Data.Fin.Tuple.Embedding
import Mathlib.Data.Finite.Card
import Mathlib.Data.Set.Card
import Mathlib.Order.CompletePartialOrder
```

but `Mathlib.Data`-whatever  looks too early to me.

Reply:
You should use `find_home!`, not `min_imports` for that (`min_imports` gives you a minimal set of `imports`, so you need *all* of them to make the proof working).

`Mathlib.GroupTheory.Index` seems to work for `exists_embedding_disjoint_range_of_add_le_ENat_card`, and maybe it also works for `restrictSurjective_of_add_le_ENatCard` if you add `import Mathlib.Data.Fin.Tuple.Embedding` (that is probably very light).

Reply:
I had seen that `Mathlib.GroupTheory.Index` works, this is what is indicated by `#find_home!`, but I don't believe this is the right place because these lemmas have nothing to do with indices of subgroups, and nobody will think of looking there.
Maybe `Mathlib.SetTheory.Cardinal.Finite`, if that works?

Reply:
You can also just create a new file if finding the right place is to annoying.

Reply:
That's what I did.